doc/test_plans/vic-machine-service/upgrade-rollback-configure.md

```diff
@@ -0,0 +1,116 @@
+
+Each API operation is implemented by a handler method. Currently, this results in duplication of code present in the `vic-machine` CLI command. This is undesirable, so this project will include work to establish patterns to avoid such duplication and refactor existing code accordingly.
+
+
+Testing Scope
+=============
+
+The primary focuses of testing will the combinatorial correctness of the re-configuration operation (including appropriate error handling for properties which may not be modified) and the end-to-end behavior of API workflows in a variety of realistic scenarios.
+
+As the API handler methods are invoked by go-swagger, testing of call routing and similar logic is out of scope.
+
+In addition to the test discussed below, careful attention should be paid to existing code that is modified as a part of the refactoring effort mentioned above. In some cases, testing coverage for that existing code may be insufficient.
+
+
+Testing Strategy
+================
+
+Unit tests
+----------
+
+Unit testing is the natural choice for verifying the behavior of the re-configuration logic; there are many potential combinations of properties that may or may not be changed as a part of a given operation and a full end-to-end test of each such combination would take prohibitively long. These tests should consider common error cases, such as ensuring proper handling of non-existent VCHs, VCHs in an invalid state, changes to properties which may not be modified, and "no-op" changes (which must be mindful of the request method used: PUT and PATCH must treat "no-op" changes to properties which may not be modified differently).
+
+Unit testing is also a natural choice for verifying any changes we make to the way that VCHs are "locked" during the upgrade process. (See https://github.com/vmware/vic/issues/6899, https://github.com/vmware/vic/issues/7083)
+
+Care must be taken when structuring the API handler logic to ensure that this functionality can be exercised without onerous mocking.
+
+
+Robot tests
+-----------
+
+Robot-based tests will be used to verify end-to-end functionality.
+
+These tests should verify functional correctness of each new API endpoint, and should as a whole exercise each parameter/feature of the endpoint, but need not execute all potential combinations of parameters/features (our end-to-end tests can focus on the most common combinations).
+
+Special attention should be paid to the interaction between combinations of operations initiated via both the CLI and API (using the API to re-configure a VCH created via the CLI, using the CLI to rollback an upgrade attempted via the API, attempts to upgrade via the CLI and API concurrently, etc.).
+
+Attention should also be paid to concurrent attempts at operations, using both the PUT and PATCH codepaths. It will be necessary to describe the desired behavior (should concurrent disjoint PATCH operations succeed?) and produce a matrix of cases to test.
```

Comment:
Might be a side discussion: do we also want to test common workflow of other vic-machine API operations together with upgrade/rollback and configure?
For example, listing VCHs when one of the VCH is under upgrade process (for ex, check if one of the VCHs in the list shows a "in upgrade" status), inspecting a VCH under upgrade, inspecting a VCH during reconfiguration, etc.
There could possibly be common customer environment workflow of combinations of different API operations. I'm not sure how likely would those occur, but as edge cases it might be valuable to consider testing?

Reply:
Added some tests like this to the concurrency section